zipline/algorithm.py

```diff
@@ -397,58 +377,54 @@ def __repr__(self):
-        """
-        Create a merged data generator using the sources attached to this
-        algorithm.
-
-        ::source_filter:: is a method that receives events in date
-        sorted order, and returns True for those events that should be
-        processed by the zipline, and False for those that should be
-        skipped.
-        """
-        if sim_params is None:
-            sim_params = self.sim_params
-
-        if self.benchmark_return_source is None:
-            if sim_params.data_frequency == 'minute' or \
-               sim_params.emission_rate == 'minute':
-                def update_time(date):
-                    return self.trading_environment.get_open_and_close(date)[1]
-            else:
-                def update_time(date):
-                    return date
-            benchmark_return_source = [
-                Event({'dt': update_time(dt),
-                       'returns': ret,
-                       'type': zipline.protocol.DATASOURCE_TYPE.BENCHMARK,
-                       'source_id': 'benchmarks'})
-                for dt, ret in
-                self.trading_environment.benchmark_returns.iteritems()
-                if dt.date() >= sim_params.period_start.date() and
-                dt.date() <= sim_params.period_end.date()
-            ]
-        else:
-            benchmark_return_source = self.benchmark_return_source
-
-        date_sorted = date_sorted_sources(*self.sources)
-
-        if source_filter:
-            date_sorted = filter(source_filter, date_sorted)
-
-        with_benchmarks = date_sorted_sources(benchmark_return_source,
-                                              date_sorted)
-
-        # Group together events with the same dt field. This depends on the
-        # events already being sorted.
-        return groupby(with_benchmarks, attrgetter('dt'))
+    def ensure_clock(self):
+        """
+        If the clock property is not set, then create one based on frequency.
+        """
+        if self.clock is None:
+            if self.sim_params.data_frequency == 'minute':
+                env = self.trading_environment
+                trading_o_and_c = env.open_and_closes.ix[
+                    self.sim_params.trading_days]
+                market_opens = trading_o_and_c['market_open'].values.astype(
+                    'datetime64[ns]').astype(np.int64)
+                market_closes = trading_o_and_c['market_close'].values.astype(
+                    'datetime64[ns]').astype(np.int64)
+                if self.sim_params.emission_rate == "daily":
+                    self.clock = MinuteSimulationClock(
+                        self.sim_params.trading_days,
+                        market_opens,
+                        market_closes,
+                        self.data_portal,
+                        env.trading_days
+                    )
+                else:
+                    self.clock = MinuteEmissionClock(
```

Comment:
In terms of names, MinuteEmission versus MinuteSimulation feel a bit too close to me to easily know the difference.